### PR TITLE
Support async KeyValuePair (de)serialization

### LINKED ISSUE #32148: The KeyValuePair converter should cache the converters for TKey and TValue
As part of https://github.com/dotnet/runtime/pull/2259, the KeyValuePair converter was identified as an area of potential perf improvement. Currently (and in 3.0\3.1) it re-enters the serializer to serialize `TKey` and `TValue` which is expensive because a Dictionary lookup occurs for each (to find the converter).

Instead, the converter should cache the converter. Caching is possible on the converter instance since the type is internal and there are no concerns about having an instance shared across different `JsonSerializerOptions` instances that may have different converters for `TKey` and `TValue`.

### PR DESCRIPTION
**Pull request by layomia**
Refactors the KeyValuePair converter to use logic for deserializing with parameterized ctors and leverage the async support (for reading and writing).

Fixes https://github.com/dotnet/runtime/issues/32148 (no longer applicable following this change).

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System.Text.Json.csproj [hunk 1]
```diff
@@ -18,10 +18,8 @@
     <NoWarn Condition="'$(TargetFramework)' == 'netstandard2.0' or $(TargetFramework.StartsWith('net4'))">$(NoWarn);nullable</NoWarn>
   </PropertyGroup>
   <ItemGroup>
-    <Compile Include="$(CommonPath)System\Runtime\CompilerServices\PreserveDependencyAttribute.cs"
-             Link="Common\System\Runtime\CompilerServices\PreserveDependencyAttribute.cs" />
-    <Compile Include="$(CommonPath)System\HexConverter.cs"
-             Link="Common\System\HexConverter.cs" />
+    <Compile Include="$(CommonPath)System\Runtime\CompilerServices\PreserveDependencyAttribute.cs" Link="Common\System\Runtime\CompilerServices\PreserveDependencyAttribute.cs" />
```
Comment (jkotas) [inline src/libraries/System.Text.Json/src/System.Text.Json.csproj:21]: Nit: Undo formatting changes?

File: src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs [hunk 2]
```diff
@@ -703,5 +689,39 @@ public static void CannotSet_DefaultIgnoreCondition_To_Always()
...
+            KeyValuePair<int, int> kvp = converter.Read(ref reader, typeToConvert, options);
+            Assert.Equal(1, kvp.Key);
+            Assert.Equal(2, kvp.Value);
+
+            // Test
+            reader = new Utf8JsonReader(bytes);
+            reader.Read();
+            try
+            {
+                converter.Read(ref reader, typeof(Dictionary<string, int>), options);
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs:716]: So this should throw InvalidOperation but today we throw NullRef?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/KeyValuePairConverter.cs [hunk 3]
```diff
@@ -3,31 +3,27 @@
...
 
 using System.Collections.Generic;
-using System.Text.Encodings.Web;
+using System.Diagnostics;
 
 namespace System.Text.Json.Serialization.Converters
 {
-    internal sealed class KeyValuePairConverter<TKey, TValue> : JsonValueConverter<KeyValuePair<TKey, TValue>>
+    internal sealed class KeyValuePairConverter<TKey, TValue> :
+        SmallObjectWithParameterizedConstructorConverter<KeyValuePair<TKey, TValue>, TKey, TValue, object, object>
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/KeyValuePairConverter.cs:11]: I'm not following. If we want to invoke the ctor on KeyValuePair then we can just call it directly -- why the need for the indirection?

File: src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs (hunk 2 above)
Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs:716]: Yes, I think it should throw InvalidOperation.

I don't think we'll null ref in all cases. I just observed one in the converter for parameterized ctors due to the converter being passed a type that it can't convert (due to wrong assumptions about metadata on `state.Current.*` ). We may have the same/other issues with other converters.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/KeyValuePairConverter.cs (hunk 3 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/KeyValuePairConverter.cs:11]: The problem is we have to duplicate all the logic for async (de)serialization (from the parameterized ctor converter) and make sure they stay in sync. The write side is relatively straightforward, but reading is more cumbersome - managing `state.Current.ObjectState`. `state.Current.PropertyState`, remembering previously parsed key/value ctor args (with `ArgumentState`) when we need to fetch more data etc.

If we had parameterized ctor support in v1 and didn't enforce that the payload have precisely 2 properties, one for the key and one for the value, we wouldn't have needed the KeyValuePair converter at all. The KeyValuePair has to exist only to do the property validation.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs [hunk 4]
```diff
@@ -440,11 +442,12 @@ private void BeginRead(ref ReadStack state, ref Utf8JsonReader reader, JsonSeria
             InitializeConstructorArgumentCaches(ref state, options);
         }
 
+        protected virtual void EndRead(ref ReadStack state) { }
+
         /// <summary>
         /// Lookup the constructor parameter given its name in the reader.
         /// </summary>
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:446]: Why was this removed?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs [hunk 5]
```diff
@@ -440,11 +442,12 @@ private void BeginRead(ref ReadStack state, ref Utf8JsonReader reader, JsonSeria
             InitializeConstructorArgumentCaches(ref state, options);
         }
 
+        protected virtual void EndRead(ref ReadStack state) { }
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:445]: Could this be abstract instead?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/KeyValuePairConverter.cs [hunk 6]
```diff
@@ -38,107 +34,66 @@ internal override void Initialize(JsonSerializerOptions options)
...
-                keySet = true;
-            }
-            else if (FoundValueProperty(propertyName, caseInsensitiveMatch))
-            {
-                reader.ReadWithVerify();
-                v = JsonSerializer.Deserialize<TValue>(ref reader, options, ref state, _valueName);
-                valueSet = true;
-            }
-            else
+            if (!state.Current.CtorArgumentState!.FoundKey &&
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/KeyValuePairConverter.cs:62]: When can `CtorArgumentState` be null? If its only when the `*WithParameterizedConstructorConverter` is not used, then maybe you could `Debug.Assert(CtorArgumentState != null)` at the beginning of the method in order to remove all the `!`.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/KeyValuePairConverter.cs (hunk 3 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/KeyValuePairConverter.cs:11]: I see. We want to support Streams properly (meaning not using read-ahead) so we are leveraging the newly added ctor support to do that.

**Comment by Jozkee** [comment]
Do we have stream-based tests with KeyValuePair?

File: src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs [hunk 7]
```diff
@@ -538,30 +548,6 @@ public static void PredefinedSerializerOptions_UnhandledDefaults(int enumValue)
...
-                ReferenceHandling = ReferenceHandling.Default,
-                WriteIndented = true,
-            };
-
-            options.Converters.Add(new JsonStringEnumConverter());
-            options.Converters.Add(new ConverterForInt32());
-
-            return options;
-        }
-
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs:564]: Is this unused code?

File: src/libraries/System.Text.Json/tests/Serialization/CollectionTests/CollectionTests.KeyValuePair.cs [hunk 8]
```diff
@@ -424,15 +424,30 @@ private class ValueNameNullPolicy : JsonNamingPolicy
...
+        public static void JsonPathIsAccurate()
+        {
+            string json = @"{""Key"":""1"",""Value"":2}";
+            JsonException ex = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<KeyValuePair<int, int>>(json));
+            Assert.Contains("$.Key", ex.ToString());
+
+            json = @"{""Key"":1,""Value"":""2""}";
+            ex = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<KeyValuePair<int, int>>(json));
+            Assert.Contains("$.Value", ex.ToString());
+        }
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/CollectionTests/CollectionTests.KeyValuePair.cs:451]: If you have a wrong property "Foo", does it show up on the path?

File: src/libraries/System.Text.Json/tests/Serialization/CollectionTests/CollectionTests.KeyValuePair.cs [hunk 9]
```diff
@@ -424,15 +424,30 @@ private class ValueNameNullPolicy : JsonNamingPolicy
...
         {
             Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<KeyValuePair<int, int>>(json));
         }
+
+        [Fact]
+        public static void JsonPathIsAccurate()
+        {
+            string json = @"{""Key"":""1"",""Value"":2}";
+            JsonException ex = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<KeyValuePair<int, int>>(json));
+            Assert.Contains("$.Key", ex.ToString());
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/CollectionTests/CollectionTests.KeyValuePair.cs:446]:
```suggestion
            Assert.Contains("$.Key", ex.ToString());
            Assert.Equal("$.Key", ex.Path);
```

Would it be good to assert both?

File: src/libraries/System.Text.Json/src/System.Text.Json.csproj (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System.Text.Json.csproj:21]: I'm not sure if we have some VS formatting rules in the repo, but I've been fighting this automatic change when saving .csproj files over multiple PRs now. The same thing happens in [this PR](https://github.com/dotnet/runtime/pull/36829/files#diff-bc2a8297e6616644e3d9970f7e8d41a9R21-R22). I think I'll let it win this time.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs (hunk 4 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:446]: The attribute does not seem to be effective on virtual methods.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs (hunk 5 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.cs:445]: I have it this way so that two derived types don't need to override it and do nothing.

File: src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs (hunk 7 above)
Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs:564]: Yes, it's unused. It should have been removed in https://github.com/dotnet/runtime/pull/36322.

File: src/libraries/System.Text.Json/tests/Serialization/CollectionTests/CollectionTests.KeyValuePair.cs (hunk 9 above)
Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/CollectionTests/CollectionTests.KeyValuePair.cs:446]: There are detailed tests for JSON path and `JsonException` are in `ExceptionTests` that provide this extra validation.

File: src/libraries/System.Text.Json/tests/Serialization/CollectionTests/CollectionTests.KeyValuePair.cs (hunk 8 above)
Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/CollectionTests/CollectionTests.KeyValuePair.cs:451]: Following this comment, yes. Thanks.

**Comment by layomia** [comment]
> Do we have stream-based tests with KeyValuePair?

Yes, see the diff in `Stream.Collections.cs`.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonValueConverterOfT.cs [hunk 10]
```diff
@@ -1,45 +0,0 @@
...
-// The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
-
-using System.Diagnostics.CodeAnalysis;
-
-namespace System.Text.Json.Serialization
-{
-    // Used for value converters that need to re-enter the serializer since it will support JsonPath
-    // and reference handling.
-    internal abstract class JsonValueConverter<T> : JsonConverter<T>
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonValueConverterOfT.cs:11]: This converter type may come back depending on work in https://github.com/dotnet/runtime/issues/36785.

**Review by steveharter — APPROVED** [review]
LGTM. Thanks
